app/src/pages/learn.tsx

```diff
@@ -0,0 +1,184 @@
+  LearningCategory,
+  LearningResource,
+} from "~/lib/validation/learn";
+
+import { learningResource } from "./api/learn";
+import { LearnCard } from "~/components/learn/learn-card";
+
+export default function Learn() {
+  const { data } = useLearningResources();
+  const { tagCategories: tagFilterOptions, typeCategories: typeFilterOptions } =
+    useLearningCategories();
+  const totalResourcesCount = useLearningResourcesTotalCount();
+
+  const [selectedCategories, setSelectedCategories] = useState<string[]>([]);
+  const [selectedTypes, setSelectedTypes] = useState<string[]>([]);
+  const [searchText, setSearchText] = useState("");
+
+  const filterData = useCallback(
+    (selectedCategories: string[], selectedTypes: string[], search: string) =>
+      data
+        .filter((resourceItem: LearningCategory) => {
+          const selectedCategoriesEmpty = selectedCategories.length === 0;
+          const selectedTypesEmpty = selectedTypes.length === 0;
+
+          const resourceTags: string[] = resourceItem.items.reduce(
+            (uniqueTags: string[], resource: LearningResource) => {
+              resource.tags.forEach((tag: string) => {
+                if (!uniqueTags.includes(tag)) {
+                  uniqueTags.push(tag);
+                }
+              });
+
+              return uniqueTags;
+            },
+            [] as string[]
+          );
+
+          if (
+            (selectedTypesEmpty || selectedTypes.includes(resourceItem.id)) &&
+            (selectedCategoriesEmpty ||
+              selectedCategories.some((cat) => resourceTags.includes(cat))) &&
+            (search === "" ||
+              resourceItem.items.some(
+                (resource: LearningResource) =>
+                  resource.title.toLowerCase().includes(search.toLowerCase()) ||
+                  resource.description
+                    .toLowerCase()
+                    .includes(search.toLowerCase())
+              ))
+          ) {
+            return true;
+          }
+          return false;
+        })
+        .map(
+          (
+            resourceItem: LearningCategory
+          ): {
+            title: string;
+            id: string;
+            items: LearningResource[];
+          } => ({
+            title: resourceItem.title,
+            id: resourceItem.id,
+            items: resourceItem.items.filter(
+              (resource: LearningResource) =>
+                (selectedCategories.length === 0 ||
+                  resource.tags.some((tag) =>
+                    selectedCategories.includes(tag)
+                  )) &&
+                (selectedTypes.length === 0 ||
+                  selectedTypes.includes(resourceItem.id)) &&
+                (search === "" ||
+                  resource.title.toLowerCase().includes(search.toLowerCase()) ||
+                  resource.description
+                    .toLowerCase()
+                    .includes(search.toLowerCase()))
+            ),
+          })
+        ),
```

Comment:
This filter function is more complicated than it should be, also not optimal as it checks for type and category selection in both the filter and map calls (which both run over every item). Here is a more simplified and optimized solution we could use here:
```typescript
        .map((resourceItem) => ({
          ...resourceItem,
          items: resourceItem.items.filter(
            (resource) =>
              (selectedCategories.length === 0 ||
                resource.tags.some((tag) =>
                  selectedCategories.includes(tag)
                )) &&
              (search.length === 0 ||
                resource.title.toLowerCase().includes(search.toLowerCase()) ||
                resource.description
                  .toLowerCase()
                  .includes(search.toLowerCase()))
          ),
        }))
        .filter(
          (resourceItem) =>
            resourceItem.items.length > 0 &&
            (selectedTypes.length === 0 ||
              selectedTypes.includes(resourceItem.id))
        ),
```